src/github.com/getlantern/detour/detour.go

```diff
@@ -43,325 +38,326 @@ import (
-		return
+// followupWrite is called by Write() if a connection's state already settled
+func (dc *Conn) followupWrite(b []byte) (n int, err error) {
+	if !dc.withValidConn(func(c conn) { c.Write(b, dc.chWrite) }) {
+		return 0, fmt.Errorf("no connection available to %s", dc.addr)
 	}
-	log.Tracef("Wrote %d bytes to %s %s", len(b), dc.addr, dc.stateDesc())
-	return
+	result := <-dc.chWrite
+	return result.n, result.err
 }
 
-// Close() implements the function from net.Conn
+// Close implements the function from net.Conn
 func (dc *Conn) Close() error {
-	log.Tracef("Closing %s connection to %s", dc.stateDesc(), dc.addr)
-	if atomic.LoadInt64(&dc.readBytes) > 0 {
-		if dc.inState(stateDetour) && wlTemporarily(dc.addr) {
-			log.Tracef("no error found till closing, add %s to permanent whitelist", dc.addr)
-			AddToWl(dc.addr, true)
-		} else if dc.inState(stateDirect) && !wlTemporarily(dc.addr) {
-			log.Tracef("no error found till closing, notify caller that %s can be dialed directly", dc.addr)
-			// just fire it, but not blocking if the chan is nil or no reader
-			select {
-			case DirectAddrCh <- dc.addr:
-			default:
-			}
-		}
+	log.Tracef("Closing connection to %s", dc.addr)
+	for len(dc.conns) > 0 {
+		conn := <-dc.conns
+		_ = conn.Close()
 	}
-	return dc.getConn().Close()
+	return nil
 }
 
-// LocalAddr() implements the function from net.Conn
+// LocalAddr implements the function from net.Conn
 func (dc *Conn) LocalAddr() net.Addr {
-	return dc.getConn().LocalAddr()
+	log.Trace("LocalAddr not implemented")
+	return nil
 }
 
-// RemoteAddr() implements the function from net.Conn
+// RemoteAddr implements the function from net.Conn
 func (dc *Conn) RemoteAddr() net.Addr {
-	return dc.getConn().RemoteAddr()
+	log.Trace("RemoteAddr not implemented")
+	return nil
 }
 
-// SetDeadline() implements the function from net.Conn
+// SetDeadline implements the function from net.Conn
 func (dc *Conn) SetDeadline(t time.Time) error {
-	if err := dc.SetReadDeadline(t); err != nil {
-		log.Debugf("Unable to set read deadline: %v", err)
-	}
-	if err := dc.SetWriteDeadline(t); err != nil {
-		log.Debugf("Unable to set write deadline: %v", err)
-	}
+	log.Trace("SetDeadline not implemented")
 	return nil
 }
 
-// SetReadDeadline() implements the function from net.Conn
+// SetReadDeadline implements the function from net.Conn
 func (dc *Conn) SetReadDeadline(t time.Time) error {
-	dc.readDeadline = t
-	if err := dc.getConn().SetReadDeadline(t); err != nil {
-		log.Debugf("Unable to set read deadline: %v", err)
-	}
+	log.Trace("SetReadDeadline not implemented")
 	return nil
 }
 
-// SetWriteDeadline() implements the function from net.Conn
+// SetWriteDeadline implements the function from net.Conn
 func (dc *Conn) SetWriteDeadline(t time.Time) error {
-	dc.writeDeadline = t
-	if err := dc.getConn().SetWriteDeadline(t); err != nil {
-		log.Debugf("Unable to set write deadline", err)
-	}
+	log.Trace("SetWriteDeadline not implemented")
```

Comment:
I just followed what `enproxy` doing. Returning an error makes the code more deterministic indeed.